examples/qtmegachatapi/megachatapplication.cpp

```diff
@@ -0,0 +1,387 @@
+
+        }, NULL);
+    }
+
+    if (newState == karere::Client::kInitHasOfflineSession ||
+            newState == karere::Client::kInitHasOnlineSession)
+    {
+        //setWindowTitle(mClient->myEmail().c_str());
+        mainWin->setWindowTitle("");
+    }
+    else
+    {
+        mainWin->setWindowTitle("");
+    }
+}
+
+
+
+void MegaChatApplication::onChatListItemUpdate(megachat::MegaChatApi* api, megachat::MegaChatListItem *item){}
+void MegaChatApplication::onChatConnectionStateUpdate(megachat::MegaChatApi* api, megachat::MegaChatHandle chatid, int newState) {}
+
+void MegaChatApplication::onChatOnlineStatusUpdate(megachat::MegaChatApi* api, megachat::MegaChatHandle userhandle, int status, bool inProgress)
+{
+    if (api->getMyUserHandle()== userhandle)
+    {
+        mainWin->ui.mOnlineStatusBtn->setText(inProgress
+            ?megachatapplication::kOnlineSymbol_InProgress
+            :megachatapplication::kOnlineSymbol_Set);
+        //ui.mOnlineStatusBtn->setStyleSheet(
+        //    kOnlineStatusBtnStyle.arg(gOnlineIndColors[api->getPresenceConfig() ? status : 0]));
+    }
+    else
+    {
+            // PENDING: update the presence for contacts
+    }
+}
+
+void MegaChatApplication::onChatPresenceConfigUpdate(megachat::MegaChatApi *api, megachat::MegaChatPresenceConfig *config)
+{
+    mainWin->ui.mOnlineStatusBtn->setText(config->isPending()
+        ?megachatapplication::kOnlineSymbol_InProgress
+        :megachatapplication::kOnlineSymbol_Set);
+  //  ui.mOnlineStatusBtn->setStyleSheet(
+   //     kOnlineStatusBtnStyle.arg(gOnlineIndColors[config->getOnlineStatus()]));
+}
+
+void MegaChatApplication::onRequestFinish(mega::MegaApi *api, mega::MegaRequest *request, MegaError *e)
+{
+    switch (request->getType())
+    {
+        case mega::MegaRequest::TYPE_LOGIN:
+            if (e->getErrorCode() == mega::MegaError::API_OK)
+            {
+               loginDlg->setState(IApp::ILoginDialog::kFetchingNodes);
+               api->fetchNodes();
+            }
+            else
+            {
+               loginDlg->setState(IApp::ILoginDialog::kBadCredentials);
+               //Request credentials again
+            }
+            break;
+
+        case mega::MegaRequest::TYPE_FETCH_NODES:
+            if (e->getErrorCode() == mega::MegaError::API_OK)
+            {
+               loginDlg->hide();
+               megaChatApi->connect();
+               mainWin->show();
+            }
+            else
+            {
+                // go back to login dialog
+            }
+            break;
+    }
+}
+
+// implementation for MegachatRequestListener
+void MegaChatApplication::onRequestFinish(megachat::MegaChatApi* megaChatApi, megachat::MegaChatRequest *request, megachat::MegaChatError* e)
+{
+    switch (request->getType())
+    {
+        case megachat::MegaChatRequest::TYPE_CONNECT:
+            if (e->getErrorCode() == mega::MegaError::API_OK)
+            {
+                KR_LOG_DEBUG("CONNECT OK");
```

Comment:
You will need to create a `MegaLogger` class and register it in `MegaApi / MegaChatApi` (one for each). Then, you can and should use that logger to post log messages, rather than using the macro (`xxx_LOG_XXX()`)

Reply:
Done